pkg/resource/providerconfig.go

```diff
@@ -46,6 +56,119 @@ func IsMissingReference(err error) bool {
+
+// EnvLookupFn looks up an environment variable.
+type EnvLookupFn func(string) string
+
+// OsEnvLookup looks up an environment variable from the OS.
+func OsEnvLookup(s string) string {
+	e, _ := os.LookupEnv(s)
+	return e
+}
+
+// ExtractEnvFn extracts credentials from an environment variable.
+func ExtractEnvFn(e EnvLookupFn) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.Env == nil {
+			return nil, errors.New(errExtractEnv)
+		}
+		return []byte(e(s.Env.Name)), nil
+	}
+
+}
+
+// ExtractFsFn extracts credentials from the filesystem.
+func ExtractFsFn(fs afero.Fs) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.Fs == nil {
+			return nil, errors.New(errExtractFs)
+		}
+		return afero.ReadFile(fs, s.Fs.Path)
+	}
+}
+
+// ExtractSecretFn extracts credentials from a Kubernetes secret.
+func ExtractSecretFn(client client.Client) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.SecretRef == nil {
+			return nil, errors.New(errExtractSecretKey)
+		}
+		secret := &corev1.Secret{}
+		if err := client.Get(ctx, types.NamespacedName{Namespace: s.SecretRef.Namespace, Name: s.SecretRef.Name}, secret); err != nil {
+			return nil, errors.Wrap(err, errGetCredentialsSecret)
+		}
+		return secret.Data[s.SecretRef.Key], nil
+	}
+}
+
+// ProviderConfigCredentialExtractor extracts credentials as specified by a ProviderConfig.
+type ProviderConfigCredentialExtractor struct {
+	handlers map[string]ExtractionHandlerFn
+}
+
+// HandlerOverride overrides or adds an extraction handler.
+type HandlerOverride struct {
+	key     string
+	handler ExtractionHandlerFn
+}
+
+// Override adds or overrides an extraction handler.
+func (h *HandlerOverride) Override(handlers map[string]ExtractionHandlerFn) {
+	handlers[h.key] = h.handler
+}
+
+// WithOverride constructs a new HandlerOverride.
+func WithOverride(key string, handler ExtractionHandlerFn) *HandlerOverride {
+	return &HandlerOverride{
+		key:     key,
+		handler: handler,
+	}
+}
+
+// NewProviderConfigCredentialExtractor returns a new credential extractor for ProviderConfigs.
+func NewProviderConfigCredentialExtractor(client client.Client, overrides ...*HandlerOverride) *ProviderConfigCredentialExtractor {
+	e := &ProviderConfigCredentialExtractor{
+		handlers: map[string]ExtractionHandlerFn{
+			"Filesystem":  ExtractFsFn(afero.NewOsFs()),
+			"Environment": ExtractEnvFn(OsEnvLookup),
+			"Secret":      ExtractSecretFn(client),
+		},
+	}
+	for _, o := range overrides {
+		o.Override(e.handlers)
+	}
+	return e
+}
```

Comment:
This seems like a bit of a roundabout way to build and return a map of handlers. Presuming we want to be able to extend the base set of handlers perhaps it would be more straightforward to go with something like:

```go
type ProviderConfigCredentialHandlers map[CredentialsSource]ExtractionHandlerFn

func NewProviderConfigCredentialHandlers(c client.Client) ProviderConfigCredentialHandlers {
	return ProviderConfigCredentialHandlers{
		CredentialsSourceFs:  ExtractFsFn(afero.NewOsFs()),
		CredentialsSourceEnv: ExtractEnvFn(OsEnvLookup),
		CredentialsSourceSecret:      ExtractSecretFn(c),
	},
}
```

For the caller this would look like:

```go
h := NewProviderConfigCredentialHandlers(client)
h[CredentialsSourceCustom] = ExtractCustomFn()
```

Or if you want to keep the ability to do it all in one call you might consider reducing the `HandlerOverride` to a simple struct - e.g.

```go
type ProviderConfigCredentialHandlers map[CredentialsSource]ExtractionHandlerFn

type Handler struct {
	Source CredentialsSource
	Fn ExtractionHandlerFn
}

func NewProviderConfigCredentialHandlers(c client.Client, h ...Handler) ProviderConfigCredentialHandlers {
	hds := ProviderConfigCredentialHandlers{
		CredentialsSourceFs:  ExtractFsFn(afero.NewOsFs()),
		CredentialsSourceEnv: ExtractEnvFn(OsEnvLookup),
		CredentialsSourceSecret:      ExtractSecretFn(c),
	},
	for _, hd := range h {
		hds[hd.Source] = hd.Fn
	}
	return hds
}
```

Reply:
I ended up with this, which feels kind of like we are wrapping a map for no reason, but I do like the idea of being able to substitute a different extractor in the future if needed:

```
// ProviderConfigCredentialExtractor extracts credentials as specified by a ProviderConfig.
type ProviderConfigCredentialExtractor map[xpv1.CredentialsSource]ExtractionHandlerFn

// NewProviderConfigCredentialExtractor returns a new credential extractor for ProviderConfigs.
func NewProviderConfigCredentialExtractor(client client.Client, c xpv1.CommonCredentialSelectors) ProviderConfigCredentialExtractor {
	return ProviderConfigCredentialExtractor{
		xpv1.CredentialsSourceFilesystem:  ExtractFsFn(afero.NewOsFs(), c),
		xpv1.CredentialsSourceEnvironment: ExtractEnvFn(os.Getenv, c),
		xpv1.CredentialsSourceSecret:      ExtractSecretFn(client, c),
	}
}

// Register adds additional handlers to the extractor.
func (p ProviderConfigCredentialExtractor) Register(source xpv1.CredentialsSource, fn ExtractionHandlerFn) {
	p[source] = fn
}

// Extract extracts credentials as specified by a ProviderConfig.
func (p ProviderConfigCredentialExtractor) Extract(ctx context.Context, source xpv1.CredentialsSource) ([]byte, error) {
	fn, ok := p[source]
	if !ok {
		return nil, errors.Errorf(errNoHandlerForSourceFmt, source)
	}
	return fn(ctx)
}
```

Thoughts?

Reply:
Been thinking about / experimenting with this more and I think I am leaning more towards just putting these common sources in a `switch` statement in something like `CommonSourceExtactor(source xpv1.CredentialsSource, client client.Client)`, then allowing consumers to use it directly if they only use the common methods, or otherwise just make it the `default` in their larger `switch`:

```
switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
	case v1beta1.CredentialSourceCustomOne:
		// handle custom source one
	case v1beta1.CredentialSourceCustomTwo:
		// handle custom source two
	default:
		resource.CommonSourceExtractor(s, client)
	}
```

They could override one of the common extractors by having a separate case for it earlier in the `switch`, or just support the extraction methods directly by using the individual ones embedded in `resource.CommonSourceExtractor`.